### ISSUE TITLE
[Bug] Weight initialization errors in Llama3 LINEAR_INIT and common FFN

### ISSUE THREAD

**Issue by Muesli1** [pytorch/torchtitan#3334]
Hi there,

While looking through the LLama3 weight init code I found (what I perceived as) two related initialization bugs affecting specifically the Llama3 model and all models using `make_ffn_config` from the common config utils.

**Bug 1: _LINEAR_INIT truncation bounds are effectively useless**

In the Llama3 model, `_LINEAR_INIT` is defined as

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/llama3/__init__.py#L43-L46 

`nn.init.trunc_normal_` has default truncation bounds of `a=-2.0` and `b=2.0`. Since `std=0.02`, this means truncation only occurs at 2.0/0.02 = 100 standard deviations from the mean, which makes the truncation completely ineffective. Same problem for `_depth_init`

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/llama3/__init__.py#L60-L64

This is inconsistent with `_output_linear_init` defined as 

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/llama3/__init__.py#L52-L57
which correctly sets explicit truncation bounds at 3 standard deviations. I suspect this is a bug and `_LINEAR_INIT` should similarly set `a=-3*0.02, b=3*0.02` (or `a=-2*0.02, b=2*0.02` to match the default 2-sigma intent).

**Bug 2: Depth-scaled init is applied to the wrong projections in make_ffn_config**

In the SwiGLU FFN, the forward pass is `w2(silu(w1(x)) * w3(x))`, which means:
- w1 is the gate projection (dim → hidden_dim)
- w3 is the value projection (dim → hidden_dim)
- w2 is the output projection (hidden_dim → dim), and the only one that writes directly back into the residual stream

as seen in 

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/common/feed_forward.py#L53-L54

The depth-scaled init should therefore be applied exclusively to `w2`. However, `make_ffn_config` as

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/common/config_utils.py#L127-L145

groups `w2` and `w3` together under `w2w3_param_init`, and this is called in the Llama3 model at

https://github.com/pytorch/torchtitan/blob/34801c0066ed07c9605516111df3b674da97910a/torchtitan/models/llama3/__init__.py#L98-L103

This means `w3` (the value projection) incorrectly receives the depth-scaled init alongside `w2`. To correct this, `w1` and `w3` should be combined into one group, leaving only `w2` with the scaled init. Note that double application of the depth-scaled init for `w3` could be quite problematic for deep models, since `w3` and `w2` are in series, so small `w3` produces small input to an already-small `w2`, making the FFN contribution to the residual stream vanishingly small at deeper layers.

I'm not sure how much influence these bugs really have in training runs, but given that torchtitan is often used as a code reference for a clean implementation, these bugs would still be nice to get fixed :)

**Comment by tianyu-l** [comment]
Thanks. I don't think weight init in torchtitan is anywhere near a "reference". As most OSS models would release their inference code but not their weight init, the default weight init we host may look more or less arbitrary.

If you have a proposal to systematically redo the init, it will be welcome and we can help review.

cc @francesco-bertolotti @garrett361

**Comment by garrett361** [comment]
> nn.init.trunc_normal_ has default truncation bounds of a=-2.0 and b=2.0. Since std=0.02, this means truncation only occurs at 2.0/0.02 = 100 standard deviations from the mean, which makes the truncation completely ineffective.

Yeah @Muesli1 , this is related to the [common complaint](https://x.com/wightmanr/status/2038740688473334075?s=20) that `a,b` aren't in units of `std`. CC @rwightman  

> The depth-scaled init should therefore be applied exclusively to w2

Yes, agreed, and I think Claude flagged this for me as well when I was looking into this recently. Depth-scaling should only be on the final linears for FFN and Attn.

Sorry @tianyu-l I haven't had a ton of time to follow up on https://github.com/pytorch/torchtitan/issues/375#issuecomment-4184756385. I did have Claude generate this terrible mess of a report (that I have not yet vetted at all) which [summarizes some init schemes](https://gist.github.com/garrett361/e2454c658ec3c78a0020be4a7b132727#code-bases-summary-table). A pretty typical pattern seems to be:
* Embedding std ~ LM Head std ~ `0.02`
* Up projection stds ~ `0.02`
* Down projection stds ~ `0.02 / sqrt(2 * n_layers)`

IMO the 0.02 factors should really be more like `~ 1 / sqrt(fan_in)` factors in most cases, but this doesn't seem typical. CC @eliebak who I was discussing this with recently.
